catchup/catchpointService.go

```diff
@@ -0,0 +1,534 @@
+	}
+
+	// download balances file.
+	ledgerFetcher := makeLedgerFetcher(cs.net, cs.ledgerAccessor, cs.log, cs)
+	attemptsCount := 0
+
+	for {
+		attemptsCount++
+
+		err = cs.ledgerAccessor.ResetStagingBalances(cs.ctx, true)
+		if err != nil {
+			if err == cs.ctx.Err() {
+				return cs.abort(err) // we want to keep it with the context error.
+			}
+			return cs.abort(fmt.Errorf("processStageLedgerDownload failed to reset staging balances : %v", err))
+		}
+		err = ledgerFetcher.downloadLedger(cs.ctx, round)
+		if err == nil {
+			break
+		}
+		if err == cs.ctx.Err() {
+			return cs.abort(err) // we want to keep it with the context error.
+		}
+
+		if attemptsCount >= maxLedgerDownloadAttempts {
+			err = fmt.Errorf("catchpoint catchup exceeded number of attempts to retrieve ledger")
+			return cs.abort(err)
+		}
+		cs.log.Warnf("unable to download ledger : %v", err)
+	}
+
+	err = cs.updateStage(ledger.CatchpointCatchupStateLastestBlockDownload)
+	if err != nil {
+		return cs.abort(fmt.Errorf("processStageLedgerDownload failed to update stage to CatchpointCatchupStateLastestBlockDownload : %v", err))
+	}
+	return nil
+}
+
+// processStageLastestBlockDownload is the third catchpoint catchup stage. It downloads the latest block and verify that against the previously downloaded ledger.
+func (cs *CatchpointCatchupService) processStageLastestBlockDownload() (err error) {
+	blockRound, err := cs.ledgerAccessor.GetCatchupBlockRound(cs.ctx)
+	if err != nil {
+		return cs.abort(fmt.Errorf("processStageLastestBlockDownload failed to retrieve catchup block round : %v", err))
+	}
+
+	fetcherFactory := MakeNetworkFetcherFactory(cs.net, 10, nil)
+	attemptsCount := 0
+	var blk *bookkeeping.Block
+	var client FetcherClient
+	// check to see if the current ledger might have this block. If so, we should try this first instead of downloading anything.
+	if ledgerBlock, err := cs.ledger.Block(blockRound); err == nil {
+		blk = &ledgerBlock
+	}
+	for {
+		attemptsCount++
+
+		if blk == nil {
+			fetcher := fetcherFactory.New()
+			blk, _, client, err = fetcher.FetchBlock(cs.ctx, blockRound)
+			if err != nil {
+				if err == cs.ctx.Err() {
+					return cs.abort(err)
+				}
+				if attemptsCount <= maxBlockDownloadAttempts {
+					// try again.
+					blk = nil
+					continue
+				}
+				return cs.abort(fmt.Errorf("processStageLastestBlockDownload failed to get block %d : %v", blockRound, err))
+			}
+			// success
+			client.Close()
+		}
+
+		// check block protocol version support.
+		if _, ok := config.Consensus[blk.BlockHeader.CurrentProtocol]; !ok {
+			cs.log.Warnf("processStageLastestBlockDownload: unsupported protocol version detected: '%v'", blk.BlockHeader.CurrentProtocol)
+			// try again.
+			blk = nil
+			continue
+		}
+
+		// check to see that the block header and the block payset aligns
+		if !blk.ContentsMatchHeader() {
+			cs.log.Warnf("processStageLastestBlockDownload: downloaded block content does not match downloaded block header")
```

Comment:
Should this and the error check above be checking `attemptsCount`?

Reply:
yes, fixed.